mypy/exprtotype.py

```diff
@@ -1,13 +1,14 @@
 """Translate an Expression to a Type value."""
 
 from mypy.nodes import (
-    Expression, NameExpr, MemberExpr, IndexExpr, TupleExpr,
+    Expression, NameExpr, MemberExpr, IndexExpr, TupleExpr, IntExpr, UnaryExpr,
     ListExpr, StrExpr, BytesExpr, UnicodeExpr, EllipsisExpr, CallExpr,
     get_member_expr_fullname
 )
 from mypy.fastparse import parse_type_comment
 from mypy.types import (
-    Type, UnboundType, TypeList, EllipsisType, AnyType, Optional, CallableArgument, TypeOfAny
+    Type, UnboundType, TypeList, EllipsisType, AnyType, Optional, CallableArgument, TypeOfAny,
+    LiteralType, RawLiteralType,
```

Comment:
I can look into doing this, but I believe we currently reference UnboundType.args in approximately 60 difference places throughout our codebase. I'm worried that propagating this change won't be cheap: we'd likely end up having to change a bunch of type signatures so we can pass through `Union[Type, RawLiteral]` and/or add a bunch of runtime checks to filter out `RawLiteral`.

(We'd also need to modify both `fastparse.TypeConverter` and `exprtotype` so they return a `Union[Type, RawLiteral]`, but I think this is a more tractable problem: we could just implement some sort of wrapper around them that reports an error and returns `AnyType` if the top-level type is `RawLiteral`)

If we're worried about RawLiterals and similar types sneaking through, I'd rather just add another layer to semantic analysis that does a final scan of all types to make sure that RawLiteral, UnboundType, and any other "bad" types are fully removed. That way, would need to scan and check each type just once, instead of in potentially multiple places whenever we manipulate `UnboundType.args`.

Or even better, we could add a new base type named `TypeLike` that `Type`, `UnboundType`, and `RawLiteral` all inherit from (along with any other fake types we don't want sneaking into the typechecking phase). `UnboundType` would then be modified so that its args contain instances of `TypeLike` rather then `Type`.

The semantic analysis layer would always work with `TypeLike`, and we'd modify `TypeAnalyzer` so it traverses over `TypeLike` and only ever returns `Type` -- this would probably let us skip the extra traversal while still giving us the guarantees we want in a typesafe way. 

The last solution would probably be just as much work, if not more, as changing `UnboundType.args` to be of type `Union[Type, RawLiteral]`, but I think it would be more principled and help us eliminate the root problem in a more fundamental way. (I'd also be willing to tackle making this change if you + the others think it's a good idea.)

Reply:
> I can look into doing this, but I believe we currently reference UnboundType.args in approximately 60 difference places throughout our codebase. I'm worried that propagating this change won't be cheap

I would recommend just trying this, if it will be soon obvious that it is too hard, then don't spend time on this. The current approach is also OK.

Reply:
FWIW, changing the type of `UnboundType.args` seems a bit ad hoc to me. We already have some `Type` subclasses such as `TypeList` that are only meaningful during semantic analysis, and I think that adding another one doesn't make things significantly worse if we are careful. Most visitors will crash if they encounter the new type subclass, so issues will likely be found soon.

Reply:
> Most visitors will crash if they encounter the new type subclass, so issues will likely be found soon.

Fair point.